File: python-practice/01-beginner/03-operators/bitwise_ops.py
"""
LESSON: Bitwise Operators
&, |, ^, ~, <<, >>, practical bit manipulation.

Run: python3 bitwise_ops.py
"""

def show_binary(name, value, bits=8):
    """Display a value in decimal, binary, and hex."""
    if value < 0:
        print(f"  {name:>12} = {value:>5}  bin: {value & 0xFF:0{bits}b}  hex: {value & 0xFF:#04x}")
    else:
        print(f"  {name:>12} = {value:>5}  bin: {value:0{bits}b}  hex: {value:#04x}")

# ===== BASIC BITWISE OPERATORS =====
print("--- Bitwise Operators ---")
a = 0b11001010  # 202
b = 0b10110101  # 181

show_binary("a", a)
show_binary("b", b)
print()

show_binary("a & b (AND)", a & b)     # both bits 1
show_binary("a | b (OR)", a | b)      # either bit 1
show_binary("a ^ b (XOR)", a ^ b)     # bits differ
show_binary("~a (NOT)", ~a)           # flip all bits
show_binary("a << 2 (LSHIFT)", a << 2)
show_binary("a >> 2 (RSHIFT)", a >> 2)

# ===== BIT SHIFT AS MULTIPLY/DIVIDE =====
print("\n--- Bit Shifting = Multiply/Divide by 2 ---")
x = 10
print(f"  {x} << 1 = {x << 1}  (x * 2)")
print(f"  {x} << 2 = {x << 2}  (x * 4)")
print(f"  {x} << 3 = {x << 3}  (x * 8)")
print(f"  {x} >> 1 = {x >> 1}  (x // 2)")

# ===== PRACTICAL: FLAGS / PERMISSIONS =====
print("\n--- Practical: Permission Flags ---")
READ    = 0b100  # 4
WRITE   = 0b010  # 2
EXECUTE = 0b001  # 1

# Set permissions
user_perms = READ | WRITE       # read + write
print(f"User perms (rw-): {user_perms:03b} = {user_perms}")

admin_perms = READ | WRITE | EXECUTE
print(f"Admin perms (rwx): {admin_perms:03b} = {admin_perms}")

# Check permission
has_read = bool(user_perms & READ)
has_exec = bool(user_perms & EXECUTE)
print(f"\nUser has READ: {has_read}")
print(f"User has EXECUTE: {has_exec}")

# Add permission
user_perms |= EXECUTE
print(f"After adding EXECUTE: {user_perms:03b}")

# Remove permission
user_perms &= ~WRITE
print(f"After removing WRITE: {user_perms:03b}")

# Toggle permission
user_perms ^= READ
print(f"After toggling READ: {user_perms:03b}")

# ===== PRACTICAL: CHECK ODD/EVEN =====
print("\n--- Practical: Odd/Even Check ---")
for n in range(8):
    parity = "odd" if n & 1 else "even"
    print(f"  {n} & 1 = {n & 1} -> {parity}")

# ===== PRACTICAL: POWER OF 2 CHECK =====
print("\n--- Practical: Power of 2 ---")
def is_power_of_2(n):
    return n > 0 and (n & (n - 1)) == 0

for n in [1, 2, 3, 4, 8, 15, 16, 32, 100, 128]:
    print(f"  {n:>4} is power of 2: {is_power_of_2(n)}")

# ===== PRACTICAL: BIT COUNTING =====
print("\n--- Practical: Count Set Bits ---")
def count_bits(n):
    return bin(n).count('1')

for n in [0, 1, 7, 15, 42, 255]:
    print(f"  {n:>4} ({n:08b}) has {count_bits(n)} set bits")

# ===== PRACTICAL: XOR TRICKS =====
print("\n--- XOR Tricks ---")

# Swap without temp
a, b = 5, 9
print(f"Before swap: a={a}, b={b}")
a ^= b
b ^= a
a ^= b
print(f"After XOR swap: a={a}, b={b}")

# Find the unique element
nums = [2, 3, 5, 3, 2]
unique = 0
for n in nums:
    unique ^= n
print(f"\nUnique in {nums}: {unique}")

File: cpp-practice/01-beginner/03-operators/bitwise_ops.cpp
/**
 * LESSON: Bitwise Operators
 * Bitwise operators work directly on the binary representation of integers.
 * Essential for low-level programming, flags, and optimization tricks.
 *
 * Compile: g++ -std=c++17 -o bitwise_ops bitwise_ops.cpp
 * Run:     ./bitwise_ops
 */
#include <iostream>
#include <bitset>  // for binary display

int main() {
    // ===== BASIC BITWISE OPERATORS =====
    std::cout << "--- Basic Bitwise Operators ---\n";
    unsigned int a = 0b1100;  // 12 in binary: 1100
    unsigned int b = 0b1010;  // 10 in binary: 1010

    std::cout << "a        = " << std::bitset<4>(a) << " (" << a << ")\n";
    std::cout << "b        = " << std::bitset<4>(b) << " (" << b << ")\n";

    // AND (&) - both bits must be 1
    std::cout << "a & b    = " << std::bitset<4>(a & b) << " (" << (a & b) << ")\n";

    // OR (|) - at least one bit must be 1
    std::cout << "a | b    = " << std::bitset<4>(a | b) << " (" << (a | b) << ")\n";

    // XOR (^) - bits must differ
    std::cout << "a ^ b    = " << std::bitset<4>(a ^ b) << " (" << (a ^ b) << ")\n";

    // NOT (~) - flip all bits
    unsigned char c = 0b00001111;
    std::cout << "~c       = " << std::bitset<8>(static_cast<unsigned char>(~c))
              << " (" << static_cast<int>(static_cast<unsigned char>(~c)) << ")\n";

    // ===== SHIFT OPERATORS =====
    std::cout << "\n--- Shift Operators ---\n";
    unsigned int val = 1;  // 0001

    // Left shift (<<) - multiply by 2^n
    std::cout << "1 << 1 = " << (val << 1) << " (multiply by 2)\n";
    std::cout << "1 << 3 = " << (val << 3) << " (multiply by 8)\n";
    std::cout << "5 << 2 = " << (5 << 2) << " (5 * 4 = 20)\n";

    // Right shift (>>) - divide by 2^n
    unsigned int big = 16;
    std::cout << "16 >> 1 = " << (big >> 1) << " (divide by 2)\n";
    std::cout << "16 >> 3 = " << (big >> 3) << " (divide by 8)\n";

    // ===== PRACTICAL: CHECK ODD/EVEN =====
    std::cout << "\n--- Check Odd/Even ---\n";
    for (int n = 0; n <= 5; n++) {
        // Last bit is 1 for odd, 0 for even
        std::cout << n << " is " << ((n & 1) ? "odd" : "even") << "\n";
    }

    // ===== PRACTICAL: SWAP WITHOUT TEMP =====
    std::cout << "\n--- Swap with XOR ---\n";
    int x = 15, y = 27;
    std::cout << "Before: x=" << x << ", y=" << y << "\n";
    x ^= y;  // x = x ^ y
    y ^= x;  // y = y ^ (x ^ y) = original x
    x ^= y;  // x = (x ^ y) ^ original x = original y
    std::cout << "After:  x=" << x << ", y=" << y << "\n";

    // ===== PRACTICAL: BIT FLAGS =====
    std::cout << "\n--- Bit Flags ---\n";
    const unsigned int READ    = 1 << 0;  // 0001 = 1
    const unsigned int WRITE   = 1 << 1;  // 0010 = 2
    const unsigned int EXECUTE = 1 << 2;  // 0100 = 4

    unsigned int permissions = 0;

    // Set flags
    permissions |= READ;              // turn on READ
    permissions |= WRITE;             // turn on WRITE
    std::cout << "Permissions: " << std::bitset<4>(permissions) << "\n";

    // Check flag
    if (permissions & READ)  std::cout << "  Has READ\n";
    if (permissions & WRITE) std::cout << "  Has WRITE\n";
    if (!(permissions & EXECUTE)) std::cout << "  No EXECUTE\n";

    // Clear a flag
    permissions &= ~WRITE;  // turn off WRITE
    std::cout << "After removing WRITE: " << std::bitset<4>(permissions) << "\n";

    // Toggle a flag
    permissions ^= EXECUTE;  // flip EXECUTE on
    std::cout << "After toggling EXECUTE: " << std::bitset<4>(permissions) << "\n";

    // ===== PRACTICAL: MASKING =====
    std::cout << "\n--- Bit Masking ---\n";
    unsigned int rgb = 0xFF5733;  // RGB color
    unsigned int red   = (rgb >> 16) & 0xFF;
    unsigned int green = (rgb >> 8) & 0xFF;
    unsigned int blue  = rgb & 0xFF;
    std::cout << "Color 0xFF5733 -> R=" << red << " G=" << green << " B=" << blue << "\n";

    // ===== POWER OF 2 CHECK =====
    std::cout << "\n--- Power of 2 Check ---\n";
    for (int n : {1, 2, 3, 4, 5, 8, 16, 15}) {
        bool isPow2 = (n > 0) && ((n & (n - 1)) == 0);
        std::cout << n << " is " << (isPow2 ? "" : "NOT ") << "a power of 2\n";
    }

    return 0;
}
